src/librustc_mir/const_eval.rs

```diff
@@ -100,15 +100,19 @@ pub fn mplace_to_const<'tcx>(
     ecx: &CompileTimeEvalContext<'_, '_, 'tcx>,
     mplace: MPlaceTy<'tcx>,
 ) -> EvalResult<'tcx, &'tcx ty::Const<'tcx>> {
-    let MemPlace { ptr, align, meta } = *mplace;
     // extract alloc-offset pair
-    assert!(meta.is_none());
-    let ptr = ptr.to_ptr()?;
+    assert!(mplace.meta.is_none());
+    let ptr = mplace.ptr.to_ptr()?;
     let alloc = ecx.memory.get(ptr.alloc_id)?;
-    assert!(alloc.align.abi() >= align.abi());
+    assert!(alloc.align.abi() >= mplace.align.abi());
     assert!(alloc.bytes.len() as u64 - ptr.offset.bytes() >= mplace.layout.size.bytes());
+    // FIXME: only clone the parts that interest us (starting at offset, going to offset + size)
     let mut alloc = alloc.clone();
-    alloc.align = align;
+    // we take `mplace.layout.align` instead of `mplace.align`
+    // as this function is essentially copying the value
+    // out of the larger allocation, so we lose all information about
+    // potential surrounding types with different alignment.
+    alloc.align = mplace.layout.align;
```

Comment:
This looks very wrong. Couldn't there still be other `ConstValue` around that refer to the same allocation?

We should never, ever change the `align` of an existing allocation. Even making a copy seems more sane, and that's already a rather insane option.^^

Reply:
oh yea, this code was broken to begin with. That's one of the uncovered bugs I was talking about.

A few commits down I'm actually creating a new Allocation from scratch instead of this wrongness. I can't make it not copy right now, that would need large changes to `ByRef` that I didn't want to do in this PR